airflow/utils/sla.py

```diff
@@ -0,0 +1,371 @@
+    """
+    Send an SLA miss email. This is the default SLA miss callback.
+    """
+    sla_miss = context["sla_miss"]
+
+    if sla_miss.sla_type == sla_miss.TASK_DURATION_EXCEEDED:
+        email_function = send_task_duration_exceeded_email
+    elif sla_miss.sla_type == sla_miss.TASK_LATE_START:
+        email_function = send_task_late_start_email
+    elif sla_miss.sla_type == sla_miss.TASK_LATE_FINISH:
+        email_function = send_task_late_finish_email
+    else:
+        log.warning("Received unexpected SLA Miss type: %s", sla_miss.sla_type)
+        return
+
+    email_to, email_subject, email_body = email_function(context)
+    send_email(email_to, email_subject, email_body)
+
+
+def describe_task_instance(ti):
+    """
+    Return a string representation of the task instance.
+    """
+    return "{dag_id}.{task_id}[{exc_date}]".format(
+        dag_id=ti.dag_id,
+        task_id=ti.task_id,
+        exc_date=ti.execution_date
+    )
+
+
+def get_sla_miss_subject(miss_type, ti):
+    """
+    Return a consistent subject line for SLA miss emails.
+    """
+    return "[airflow] [SLA] {miss_type} on {task_instance}".format(
+        miss_type=miss_type,
+        task_instance=describe_task_instance(ti)
+    )
+
+
+def get_subscribers(tasks):
+    """
+    Return a list of unique emails from a list of tasks.
+    """
+    def _yield_subscribers(tasks):
+        for t in tasks:
+            if t.email:
+                if isinstance(t.email, basestring):
+                    yield t.email
+                else:
+                    for e in t.email:
+                        yield e
+    return list(set(_yield_subscribers(tasks)))
+
+
+@provide_session
+def get_blocked_task_instances(task_instance, session=None):
+    """
+    Given a task instance, return task instances that may currently
+    be blocked by it.
+    """
+    dag = task_instance.task.dag
+
+    TI = airflow.models.TaskInstance
+    downstream_tasks = task_instance.get_flat_relatives(upstream=False)
+
+    # The intent is to capture states that indicate that work was never started
+    # on a task, presumably because this task not achieving its SLA prevented
+    # the downstream task from ever successfully starting.
+
+    # It is possible for an upstream task to cause a downstream task to *fail*,
+    # like if it never produces a required artifact. But in a well-behaved DAG
+    # where dependencies are encoded properly, this shouldn't happen.
+    blocked_states = (
+        State.UPSTREAM_FAILED,
+        State.SCHEDULED,
+        State.QUEUED,
+        State.NONE,
+    )
+
+    qry = (
+        session
+        .query(TI)
+        .filter(TI.dag_id == dag.dag_id)
+        .filter(TI.task_id.in_(list(t.task_id for t in downstream_tasks)))
+        .filter(TI.execution_date == task_instance.execution_date)
+        .filter(TI.state.in_(blocked_states))
+        .all()
+    )
+    return qry
+
+
+def send_task_duration_exceeded_email(context):
```

Comment:
Please also add documentation to these functions.

Reply:
Added some documentation to these functions and improved it generally across this module